runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -2679,6 +2679,32 @@ bool isConditionCodeSetForCompare(TR::Node *node, bool *jumpOnOppositeCondition)
    return false;
    }
 
+void setImplicitNULLCHKExceptionInfo(TR::Node *node, TR::CodeGenerator *cg, TR::Compilation *comp)
+   {
+   if (!node->hasFoldedImplicitNULLCHK())
+      return;
+
+   // Correctly set the implicit exception point
+   //
+   // The compare analyzer may have generated a secondary load to
+   // resolved a compressed pointer and incorrectly sets the exception point.
+   // If that is case, correctly set exception point on the comparison that
+   // throws the null pointer exception.
+   TR::Instruction *cmp = cg->getAppendInstruction()->getPrev();
+   bool isCmpWithCompressedPtr = comp->useCompressedPointers() && cmp->getOpCodeValue() == TR::InstOpCode::CMP4MemReg;
+   if (isCmpWithCompressedPtr)
+      cg->setImplicitExceptionPoint(cmp);
+
+   TR::Instruction *faultingInstruction = cg->getImplicitExceptionPoint();
+   if (faultingInstruction)
+      {
```

Comment:
I wonder if the logic you have above to look for the compare instruction would be better to go here.  Once you know there is a `faultingInstruction`, if the compare instruction is a memory form instruction then it should become the implicit exception instruction if it already isn't.